File: app/src/main/java/com/garymcgowan/moviepedia/view/search/MovieListAdapter.java
package com.garymcgowan.moviepedia.view.search;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.garymcgowan.moviepedia.R;
import com.garymcgowan.moviepedia.model.Movie;
import com.garymcgowan.moviepedia.view.details.MovieDetailsActivity;
import com.squareup.picasso.Picasso;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MovieListAdapter extends RecyclerView.Adapter<MovieListAdapter.ViewHolder> {

    private final List<Movie> movieList;

    public MovieListAdapter(List<Movie> items) {
        movieList = items;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_movie, parent, false);
        ViewHolder vh = new ViewHolder(view);

//        RxView.clicks(view)
//                .map(v -> vh.getAdapterPosition())
//                .subscribe(onClickSubject);

        return vh;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Movie currentMovie = movieList.get(position);
//set movie title
        holder.titleTextView.setText(currentMovie.getTitleYear());

        //load image with Picasso
        Picasso.with(holder.mView.getContext())
                .load(currentMovie.getPosterURL())
                .placeholder(R.color.imagePlaceholder)
                .error(R.color.imagePlaceholder)
                .into(holder.posterImageView);

        holder.mView.setOnClickListener(v -> {
            //launch details intent
            Context context = v.getContext();
            Intent intent = new Intent(context, MovieDetailsActivity.class);
            intent.putExtra(MovieDetailsActivity.ARG_ITEM_ID, currentMovie.getImdbID());
            intent.putExtra(MovieDetailsActivity.ARG_ITEM_TITLE, currentMovie.getTitle());

            context.startActivity(intent);

        });
    }

    @Override
    public int getItemCount() {
        if (movieList == null) {
            return 0;
        }
        return movieList.size();
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        final View mView;

        @BindView(R.id.titleTextView) TextView titleTextView;
        @BindView(R.id.posterImageView) ImageView posterImageView;

        public ViewHolder(View view) {
            super(view);
            mView = view;
            ButterKnife.bind(this, view);
        }
    }
}

File: app/src/main/java/com/garymcgowan/moviepedia/view/search/MovieListAdapter.kt
package com.garymcgowan.moviepedia.view.search

import android.content.Intent
import android.os.Bundle
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.navigation.Navigation
import com.garymcgowan.moviepedia.R
import com.garymcgowan.moviepedia.model.Movie
import com.garymcgowan.moviepedia.view.details.MovieDetailsFragment
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.item_movie.view.*

class MovieListAdapter(private val movieList: List<Movie>?, private val favCallback: (Movie) -> Unit) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_movie, parent, false)

        //        RxView.clicks(view)
        //                .map(v -> vh.getAdapterPosition())
        //                .subscribe(onClickSubject);

        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {

        val currentMovie = movieList!![position]
        when (holder) {
            is ViewHolder -> {
                //set movie title
                holder.titleTextView.text = currentMovie.titleYear()

                //load image with Picasso
                Picasso.with(holder.parentView.context).load(currentMovie.posterURL).placeholder(R.color.imagePlaceholder).error(R.color.imagePlaceholder).into(holder.posterImageView)

                holder.parentView.setOnClickListener { v ->
                    Navigation.findNavController(v)
                            .navigate(
                                    R.id.navigation_details,
                                    MovieDetailsFragment.safeArgs(currentMovie.imdbID, currentMovie.title)
                    )
                }

                holder.favouriteButton.setOnClickListener {
                    favCallback.invoke(currentMovie)
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return movieList?.size ?: 0
    }

    internal class ViewHolder(val parentView: View) : RecyclerView.ViewHolder(parentView) {
        var titleTextView: TextView = parentView.titleTextView
        var posterImageView: ImageView = parentView.posterImageView
        var favouriteButton: View = parentView.favourite_button
    }
}
